protocol/lavasession/router_key.go

```diff
@@ -4,34 +4,78 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+
+	"github.com/lavanet/lava/v3/utils/lavaslices"
 )
 
 const (
-	sep            = "|"
-	methodRouteSep = "method-route:"
+	RouterKeySeparator = "|"
+	methodRouteSep     = "method-route:"
+	internalPathSep    = "internal-path:"
 )
 
-type RouterKey string
-
-func (rk *RouterKey) ApplyMethodsRoute(routeNum int) RouterKey {
-	additionalPath := strconv.FormatInt(int64(routeNum), 10)
-	return RouterKey(string(*rk) + methodRouteSep + additionalPath)
+type RouterKey struct {
+	methodsRouteUniqueKey int
+	uniqueExtensions      []string
+	internalPath          string
 }
 
 func NewRouterKey(extensions []string) RouterKey {
+	routerKey := RouterKey{}
+	routerKey.SetExtensions(extensions)
+	return routerKey
+}
+
+func GetEmptyRouterKey() RouterKey {
+	return NewRouterKey([]string{})
+}
+
+func (rk *RouterKey) SetExtensions(extensions []string) {
 	// make sure addons have no repetitions
-	uniqueExtensions := map[string]struct{}{}
-	for _, extension := range extensions {
-		uniqueExtensions[extension] = struct{}{}
+	uniqueExtensions := map[string]struct{}{} // init with the empty extension
+	if len(extensions) == 0 {
+		uniqueExtensions[""] = struct{}{}
+	} else {
+		for _, extension := range extensions {
+			uniqueExtensions[extension] = struct{}{}
+		}
 	}
+
 	uniqueExtensionsSlice := []string{}
 	for addon := range uniqueExtensions { // we are sorting this anyway so we don't have to keep order
 		uniqueExtensionsSlice = append(uniqueExtensionsSlice, addon)
 	}
+
 	sort.Strings(uniqueExtensionsSlice)
-	return RouterKey(sep + strings.Join(uniqueExtensionsSlice, sep) + sep)
+
+	rk.uniqueExtensions = uniqueExtensionsSlice
 }
 
-func GetEmptyRouterKey() RouterKey {
-	return NewRouterKey([]string{})
+func (rk *RouterKey) ApplyMethodsRoute(routeNum int) {
+	rk.methodsRouteUniqueKey = routeNum
+}
+
+func (rk *RouterKey) ApplyInternalPath(internalPath string) {
+	rk.internalPath = internalPath
+}
+
+func (rk RouterKey) HasExtension(extension string) bool {
+	return lavaslices.Contains(rk.uniqueExtensions, extension)
+}
+
+func (rk RouterKey) String() string {
```

Comment:
This one doesn't change, I dont want us to calculate it every time string is called. it would be better to just store the key string in the struct once and return it. 
if rk.key != "" {
    return rk.key 
}

else calculate it and store.

Reply:
It is dangerous because we might add some logic that will actually change it in the run, and then we will get unexpected behavior. There is currently only one place in the code that creates the router key and call its String function more than once. I fixed it there in this commit: b3982570d647634e91812c6cb049ca497f0d30de